### ISSUE TITLE
Panel edit mode in Serena

### ISSUE THREAD

**Issue by Cos-mo** [linuxmint/Cinnamon#6140]
I just came across in Roadmap -> Triaged.md about this, noted under "Cannot reproduce":

> If in panel edit mode right clicking the panel doesn’t do anything. This means, that you cannot turn off panel edit mode.

Actually I reported this in the beta testing blog and the issue is still present. I give some more info about it:

Test environment: VirtualBox 5.0.30; guest additions installed. The issue is independent, if 3D support is activated in the settings for the VM or not.
Used: Mint 18.1 64 bit, beta build (formerly) and fresh installed final release.

As long as the panel is not moved from the default (bottom) position, all works as expected.

As soon as I move the panel to any other place (top or side) I can enter the panel edit mode via right clicking the panel. But if I right click the panel again to leave panel edit mode nothing happens. I can click as often as I want.
This is not an issue of the pointing device because right clicking the desktop works at once. Actually right clicking the desktop and maneuvering to the panel settings dialog is the only possible way to get out of this situation.
Note: If I am not in panel edit mode the right clicks on the panel work as expected.

Even if I move the panel back to the bottom, activating panel edit mode and right clicking in panel edit mode (for leaving it) does nothing.

Further finding:
If the panel is placed at the bottom or top the problem can get cured by restarting Cinnamon (ctrl-alt-esc). Once Cinnamon has been restarted, I can enter and leave panel edit mode via right clicking the panel as often as I wish (as long as I do not move the panel again).
If the panel is placed at the left or the right restarting Cinnamon helps so far, as this quits the panel edit mode. But if I now enter pane edit mode again, I still cannot leave it via right clicking the panel.

glass.log does not contain anything regarding the bug.

All level 1 to 3 updates until today have been applied. After that also all level 4 & 5 have been applied. This did not change anything for this issue.

Hope this helps to make the bug reproducible.

**Comment by brownsr** [comment]
Hi Cos-mo.  Could you see if there is any link to the screensaver settings - turn off the screensaver/locker and see if the same issue is there ?  You mentioned that restarting Cinnamon quits panel edit mode.  Yes, it does, that's by design.

**Comment by Cos-mo** [comment]
Thank you for replying, brownsr.

I have the screensaver in general disabled on all my virtual machines, as they do not make any sense in this environment, in case of a test system especially (and this jumping new thing makes me nervous). So we can exclude this.

I do know, that restarting Cinnamon disables panel edit mode by design. I did mention this only for clarification, what I did and what happened here.

One addition: If I boot the live system in the virtual machine the bug is reproducible there; tested with moving the panel to the top. So I can clearly exclude any personal settings in the user account. (In the test system I do nearly none personal settings at all except disabling the screensaver.)

**Comment by brownsr** [comment]
Thanks for your accurate details on how to reproduce this.  I can reproduce it on virtualbox 5.0.24 in 'live' mode.  I can also offer you a workaround (which is further information that might lead to the source of the issue).  It only appears to happen if the central box in a panel is empty.  If you drag any applet in there then right click works fine. That includes the window list, even if empty because no applications are running, and even a spacer applet [dragging these in can be a bit fiddly] if you want the centre of the panel to look unoccupied

I do see a couple of errors at various times in the .xsession-errors log which might be related:

meta_screen_get_monitor_geometry  assertion 'monitor >=0 && monitor < screen -> n_monitor_infos failed"

cinnamon_tray_manager_manage_stage:  assertion manager->priv->stage == NULL  failed

I'll see if I can reproduce it on bare metal now, to see if it is just related to running in virtualbox, or to live mode.

**Comment by brownsr** [comment]
@Cos-mo Would you be able to turn off menu animation (settings, effects, effects on menus) to see if that makes a difference for you ?

**Comment by Cos-mo** [comment]
@brownsr 
Thank you once more. I also thank you for the workaround, although this was not my intention. I also described 2 workarounds (panel settings dialog and restarting Cinnamon) which both have the advantage, that I and not the system decide, if and where I place an applet.

My intention for this report was and is to solve the unsatisfactory situation, that an undoubtedly existing bug (special thank for the confirmation) did not get fixed and that I only by random found the note in roadmap, that it had been noted as not reproducible. So my conclusion was, that it is my task to describe the problem and the steps and conditions for reproducing as detailed, as I can imagine. (The other unsatisfactory point is, that I got after my first report never a reply about the inability to reproduce it; this makes bug hunting and reporting frustrating.)

I just found during writing the above your last reply and did as you suggested: I confirm, that turning off the menu effects lets me open the context menu in panel edit mode. If I turn them on again the context menu (in edit mode) is again not more reachable. Regardless of the position of the panel. It seems, that you are on the right track.

**Comment by brownsr** [comment]
Thanks for checking that @Cos-mo

**Comment by Cos-mo** [comment]
I did some more investigation into the issue. For preventing misunderstandings: The following tests have been done with effects enabled.

The "trick" to place an applet into the middle section of the panel is not the whole story in case of vertical panels. There must be at least one applet in every section (left, middle, right) of the panel.

Reproducible steps:
Starting from the default panel layout (one panel at the bottom, only default applets) I added a second panel at the right, then restarted Cinnamon. Now I dragged an applet from the bottom panel to the middle section of the right panel. Trying to stop panel edit mode via right clicking the panel is impossible, no context menu appears, regardless if I right click the right or the bottom panel. By dragging other applets from the bottom panel to the right panel I found, that there must be (as said) at least one applet in every section of the right panel.
If I create a second panel on the top instead to the right this problem does not happen. In the top panel no applet is needed to get the right click menu in panel edit mode.

To sum up:
The bug appears if a vertical panel exists and there is not at least one applet in every section of the vertical panel. This happens regardless, if the vertical panel is the only panel (panel moved from bottom to the side) or if the vertical panel has been added as a second panel (in addition to the bottom).
The bug also appears, if the bottom panel gets **moved** to the top and you do **not** restart Cinnamon afterwards. If the top panel gets **added** as a second panel (in addition to the bottom) (without restarting Cinnamon) nothing bad happens, even if there is no applet in the top panel at all. A counter check with Sarah reveals, that you can move the panel from the bottom to the top without the need to restart Cinnamon. So we have a clear regression.
Note: Also in the case, when you move the bottom panel to another position and then back to the bottom you have to restart Cinnamon at first to prevent the bug happening.

This gives, that not only the vertical panels are affected by the bug, but in vertical panels the bug appears also in a situation, where it does not happen in horizontal panels.

**Comment by brownsr** [comment]
@Cos-mo We tracked the problem down to excessive allocation looping in edit mode.  This affected both vertical and horizontal panels.  This in turn affected the right click menu animation, which could cause sticking in edit mode if the animation failed to complete.  I suspect it showed up more readily in virtualbox because only one core might have been enabled and/or it was more sensitive to the heavy CPU use.  I found the same symptoms on a low-power touch laptop, whereas my more powerful development laptop ran fine.  I've pushed a fix.  Releasing it is subject to review by the wider team.
